feat(dispute): add rootClaimByChainId(uint256) for L2 chain ID lookup

stevennevins opened the pull request:
Add `rootClaimByChainId` function to IDisputeGame interface that accepts an L2 chain ID parameter. This prepares the dispute game contracts for interop support where claims will need to be retrieved by chain ID.

Pre-interop games (FaultDisputeGame, FaultDisputeGameV2, OPSuccinctFaultDisputeGame) return the existing root claim regardless of chain ID. SuperFaultDisputeGame reverts with
SuperRootPreimageNotSupported until super root preimage support is added.

<!--
Contributions welcome! See https://github.com/ethereum-optimism/.github/blob/master/CONTRIBUTING.md
-->

**Description**

<!--
A clear and concise description of the features you're adding in this pull request.
-->

**Tests**

<!--
Please describe any tests you've added. If you've added no tests, or left important behavior untested, please explain why not.
-->

**Additional context**

<!--
Add any other context about the problem you're solving.
-->

**Metadata**

Completes https://github.com/ethereum-optimism/optimism/issues/18516

stevennevins committed:
feat(dispute): add rootClaimByChainId for L2 chain ID lookup

Add rootClaimByChainId(uint256) to IDisputeGame interface for interop
preparation. Pre-interop games return the existing rootClaim regardless
of chain ID. SuperFaultDisputeGame reverts with SuperRootPreimageNotSupported
until super root preimage support is added.

Renamed from rootClaim(uint256) to avoid function overload ambiguity
with abi.encodeCall (semgrep sol-style-use-abi-encodecall rule).

stevennevins committed:
chore: resolve ci issues

stevennevins committed:
fix: restore accidentally deleted snapshot files

Restores abi_loader.go, abi_loader_test.go, and state-diff/.gitkeep
that were accidentally removed in previous commit.

stevennevins committed:
fix: update StandardValidator version assertions for dispute games

Update permissionedDisputeGameVersion() to match new dispute game
versions (V1: 1.9.0, V2: 2.3.0) and bump validator to 2.3.0.

stevennevins committed:
fix: rename test to follow 4-part revert naming convention

Inphi commented on packages/contracts-bedrock/interfaces/dispute/v2/IFaultDisputeGameV2.sol:
```diff
@@ -114,6 +114,7 @@ interface IFaultDisputeGameV2 is IDisputeGame {
     function move(Claim _disputed, uint256 _challengeIndex, Claim _claim, bool _isAttack) external payable;
     function normalModeCredit(address) external view returns (uint256);
     function refundModeCredit(address) external view returns (uint256);
+    function rootClaimByChainId(uint256) external pure returns (Claim rootClaim_);
```
Did you mean `rootClaim(uint256)`. The current name is OK, just wasn't sure if this is a typo given the PR description.

stevennevins replied:
there was an issue with the current vm.mock call usage and overloaded names iirc. the naming here was intention. Didn't update the PR description to align. will address that now

Inphi commented on packages/contracts-bedrock/src/L1/OPContractsManagerStandardValidator.sol:
```diff
@@ -220,9 +220,9 @@ contract OPContractsManagerStandardValidator is ISemver {
     function permissionedDisputeGameVersion() public view returns (string memory) {
         if (DevFeatures.isDevFeatureEnabled(devFeatureBitmap, DevFeatures.DEPLOY_V2_DISPUTE_GAMES)) {
-            return "2.2.0";
+            return "2.3.0";
```
this needs a rebase on develop. We no longer hardcode DG versions in std validator

Inphi commented on packages/contracts-bedrock/src/dispute/FaultDisputeGame.sol:
```diff
@@ -865,6 +865,13 @@ contract FaultDisputeGame is Clone, ISemver {
+    /// @dev For pre-interop games, returns the root claim regardless of chain ID.
+    /// @return rootClaim_ The root claim of the DisputeGame.
+    function rootClaimByChainId(uint256) public pure returns (Claim rootClaim_) {
+        rootClaim_ = rootClaim();
```
what do you think about reverting if the input chain ID isn't known? As a defensive check against a misconfigured bridge or dispute game?

stevennevins replied:
sgtm

stevennevins committed:
chore: reduce diff by going back to old selector usage

stevennevins committed:
fix: undeclared identifier error

stevennevins committed:
fix: undeclared identifier error

stevennevins committed:
feat: add defensive check on rootClaimByChainId

stevennevins committed:
fix: undeclared identifier error

stevennevins committed:
feat: add defensive check on rootClaimByChainId for V2

Revert with UnknownChainId if input chain ID doesn't match l2ChainId().
Prevents misconfigured bridge from silently returning wrong data.

stevennevins committed:
fix: interfaces

Inphi commented on packages/contracts-bedrock/interfaces/dispute/IDisputeGame.sol:
```diff
@@ -13,6 +13,7 @@ interface IDisputeGame is IInitializable {
     function gameType() external view returns (GameType gameType_);
     function gameCreator() external pure returns (address creator_);
     function rootClaim() external pure returns (Claim rootClaim_);
+    function rootClaimByChainId(uint256 _chainId) external view returns (Claim rootClaim_);
```
nit, for consistency with `rootClaim` as it'll be odd if it needed a different modifier.
```suggestion
    function rootClaimByChainId(uint256 _chainId) external pure returns (Claim rootClaim_);
```

Inphi reviewed (approved):
Looks good.

stevennevins committed:
fix: consistent function modifier

stevennevins committed:
fix: use pure and update snapshots

stevennevins committed:
test: add rootClaimByChainId tests for permissioned dispute games

Add fuzz tests for rootClaimByChainId function to PermissionedDisputeGame
and SuperPermissionedDisputeGame test contracts.

opgitgovernance commented:
This pr has been automatically marked as stale and will be closed in 5 days if no updates

ajsutton reviewed (approved):
LGTM.

stevennevins committed:
fix: remove duplicate declarations from rebase
